### PR TITLE
Add Dockerfile for server & studio

### LINKED ISSUE #184: Hugegraph 能否支持在docker中运行
### Expected behavior 期望表现
{type something here...}

### Actual behavior 实际表现
{type something here...}

### Steps to reproduce the problem 复现步骤
1. {step 1}
2. {step 2}
3. {step 3}

#### Vertex/Edge summary 数据量
- loaded vertices amount: {like 10 million}
- loaded edges amount: {like 20 million}
- loaded time: {like 200s}

#### Vertex/Edge example 数据示例
{type something here...}

#### Schema(VertexLabel, EdgeLabel, IndexLabel) 元数据结构
{type something here...}

### Specifications of environment 环境信息
- hugegraph version: {like v0.7.4}
- operating system: {like centos 7.4, 32 CPUs, 64G RAM}
- hugegraph backend: {like cassandra 3.10, cluster with 20 nodes, 3 x 1TB HDD disk each node}

### PR DESCRIPTION
**Pull request by imbajin**
Fix #184 #840  #989 :

0. Use rocksdb as the default storage engine
1. Add hosts to speed up image build (comment by default)
2. Modify some configs to enable the server to run better in container
3. Because Dockerfile doesn't allow multiple startup commands, currently only the server will be automatically started after the container is started. If you want to start studio, enter the container to start it.

Here are some simple using commands:
```bash
# 0. download image
docker pull imbajin:hugegraph

# 1. start container
docker run -itd imbajin:hugegraph  # quick start

# (Optional configs)
docker run -itd --name=graph -p 18080:8080 -p 18088:8088 imbajin:hugegraph

# 2. (Optional) enter the container
docker exec -it graph /bin/bash
```

**TODOs**:
1. Build from the latest code instead of tar package.
2. Support [hubble](https://github.com/hugegraph/hugegraph-hubble) studio
3. Add & split studio/loader into separate images, use `docker-compose` to manage them.
4. Maybe have better way to modify the configs to satisfied the container (like add a option for running in the foreground)

After the above completed, it can provides a one-click **pure** test environment for graph

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    lsof \
+    g++ \
+    gcc \
+    openjdk-8-jdk \ 
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/* 
+
+# 2. Init HugeGraph Sever (speed up)
+RUN set -e \
+    && echo "192.30.253.112 github.com\n151.101.44.249 github.global.ssl.fastly.net" >> /etc/hosts \
```
Comment (javeme) [inline Dockerfile]: is the ip stable?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    && mkdir -p /root/server \
+    && curl -L -S ${PKG_URL}/hugegraph/releases/download/v0.10.4/hugegraph-0.10.4.tar.gz -o /root/server.tar.gz \
+    && tar xzf /root/server.tar.gz --strip-components 1 -C /root/server \
+    && rm /root/server.tar.gz \
+    && cd /root/server/ \
+    && sed -i "s/^restserver.url.*$/restserver.url=http:\/\/0.0.0.0:8080/g" ./conf/rest-server.properties \
+    && sed -n '63p' ./bin/start-hugegraph.sh | grep "&" > /dev/null && sed -i 63{s/\&$/#/g} ./bin/start-hugegraph.sh \
+    && sed -n '74p' ./bin/start-hugegraph.sh | grep "exit" > /dev/null && sed -i 74{s/^/#/g} ./bin/start-hugegraph.sh \
+    && ./bin/init-store.sh
```
Comment (javeme) [inline Dockerfile]: add a shell install-server.sh?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+# 3. Prepare for HugeGraph Studio
+RUN set -e \
+    && echo "192.30.253.112 github.com\n151.101.44.249 github.global.ssl.fastly.net" >> /etc/hosts \
+    && mkdir -p /root/studio \
+    && curl -L -S ${PKG_URL}/hugegraph-studio/releases/download/v0.10.0/hugegraph-studio-0.10.0.tar.gz -o /root/studio.tar.gz \
```
Comment (javeme) [inline Dockerfile]: seems no studio tar when building server

File: Dockerfile (hunk 2 above)
Comment (imbajin) [reply Dockerfile]: Need test it:
if there is a file named `start.sh`
```
echo "old msg"
```
1. `RUN start.sh` 
2. `RUN echo "old msg" && echo "new msg"
If we use a shell file to change the code, would dockerfile think this is a repeated step and skip the rebuild?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,46 @@
+FROM ubuntu:xenial
```
[the remark this answers is not in the document]
Comment (imbajin) [reply Dockerfile]: We can only write a basically similar Dockerfile for CentOS. (Generally use `Debian/Ubuntu`)
Generally, I haven't seen the official image provide two types of Dockerfile, 
Maybe we can see the opinions of other community users.

And here are some refers:
1. [Cassandra_Dockerfile](https://github.com/docker-library/cassandra/blob/e5b7ed497ff74587560682dd4c6a2e9b01f818d9/3.11/Dockerfile)
2. [Nginx_Dockerfile](https://github.com/nginxinc/docker-nginx/blob/5971de30c487356d5d2a2e1a79e02b2612f9a72f/mainline/buster/Dockerfile)
3. [Other Official Dockerfile](https://hub.docker.com/search?q=&type=image&image_filter=official)

[the remark this answers is not in the document]
Comment (javeme) [reply Dockerfile]: OK

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+# 1. Install needed dependencies of GraphServer & RocksDB
+RUN set -x \
+    && apt-get -q update \
+    && apt-get -q install -y --no-install-recommends --no-install-suggests \
+    curl \
+    lsof \
+    g++ \
+    gcc \
+    openjdk-8-jdk \
```
Comment (javeme) [inline Dockerfile]: merge the 11-15 lines, and align with apt-get

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    && cd /root/hugegraph-server/ \
+    && sed -i "s/^restserver.url.*$/restserver.url=http:\/\/0.0.0.0:8080/g" ./conf/rest-server.properties \
+    && sed -n '63p' ./bin/start-hugegraph.sh | grep "&" > /dev/null && sed -i 63{s/\&$/#/g} ./bin/start-hugegraph.sh \
+    && sed -n '74p' ./bin/start-hugegraph.sh | grep "exit" > /dev/null && sed -i 74{s/^/#/g} ./bin/start-hugegraph.sh \
+    && ./bin/init-store.sh
+
+# 3. Prepare for HugeGraph Studio
+ENV STUDIO_VERSION 0.10.0
+RUN set -e \
+    && echo "192.30.253.112 github.com\n151.101.44.249 github.global.ssl.fastly.net" >> /etc/hosts \
```
Comment (javeme) [inline Dockerfile]: remove it since added when building server

Comment (imbajin) [reply Dockerfile]: Each layer of docker image use a separate network layer.
So when leave current layer(step), other step cannot reuse it

And other ways are troublesome to solve it, like [this](https://stackoverflow.com/questions/38302867/how-to-update-etc-hosts-file-in-docker-image-during-docker-build/38920219#38920219)

Comment (javeme) [reply Dockerfile]: Get it

File: Dockerfile (hunk 3 above)
Comment (javeme) [reply Dockerfile]: we may need to also trigger docker image building when studio release to resolve it

**Comment by imbajin** [comment]
Also update default value for auth for #706

**Comment by javeme** [comment]
> Also update default values for auth for #706

There are indeed some errors, please update default value to ConfigAuthenticator:
`auth.authenticator=com.baidu.hugegraph.auth.ConfigAuthenticator`

**Comment by javeme** [comment]
@imbajin Can you split the commits not related to docker?

**Comment by imbajin** [comment]
> @imbajin Can you split the commits not related to docker?

OK, maybe it's better to update all the configuration files & shells separately later

File: hugegraph-dist/src/assembly/static/conf/hugegraph.properties [hunk 7]
```diff
@@ -42,7 +42,7 @@ cassandra.password=
 # hbase backend config
 #hbase.hosts=localhost
 #hbase.port=2181
-#hbase.znode_parent=/hbsae
+#hbase.znode_parent=/hbase
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/conf/hugegraph.properties:45]: also move it to independent commit

**Review by javeme — APPROVED** [review]

**Review by Linary — APPROVED** [review]

**Review by zhoney — APPROVED** [review]
